### PR TITLE
[OTPlugin] Per-channel OpenTelemetry plugin

### PR DESCRIPTION
**Pull request by yijiem**

### CODE REVIEW

File: include/grpcpp/ext/otel_plugin.h [hunk 1]
```diff
@@ -142,6 +152,10 @@ class OpenTelemetryPluginBuilder {
   /// Registers a global plugin that acts on all channels and servers running on
   /// the process.
   absl::Status BuildAndRegisterGlobal();
+  /// Builds an open telemetry plugin, returns the plugin object when succeeded
+  /// or an error status when failed.
+  GRPC_MUST_USE_RESULT absl::StatusOr<std::shared_ptr<OpenTelemetryPlugin>>
```
Comment (yashykt) [inline include/grpcpp/ext/otel_plugin.h:157]: any special reason why this returns a shared_ptr?

File: include/grpcpp/ext/otel_plugin.h [hunk 2]
```diff
@@ -26,24 +26,31 @@
...
 
 class OpenTelemetryPluginOption {
  public:
   virtual ~OpenTelemetryPluginOption() = default;
 };
 
+class OpenTelemetryPlugin {
+ public:
+  // Adds this OpenTelemetryPlugin to the channel args \a args.
+  virtual void AddToChannelArguments(grpc::ChannelArguments* args) = 0;
```
Comment (yashykt) [inline include/grpcpp/ext/otel_plugin.h:51]: probably should not be virtual?

File: include/grpcpp/ext/otel_plugin.h [hunk 3]
```diff
@@ -113,8 +120,8 @@ class OpenTelemetryPluginBuilder {
   /// If set, \a generic_method_attribute_filter is called per call with a
   /// generic method type to decide whether to record the method name or to
   /// replace it with "other". Non-generic or pre-registered methods remain
-  /// unaffected. If not set, by default, generic method names are replaced with
-  /// "other" when recording metrics.
+  /// unaffectePd. If not set, by default, generic method names are replaced
```
Comment (yashykt) [inline include/grpcpp/ext/otel_plugin.h:123]: please fix

**Comment by yashykt** [comment]
Let's make sure we're passing this channel arg to the rls channel and the xds client channel as well

File: include/grpcpp/ext/otel_plugin.h [hunk 4]
```diff
@@ -26,24 +26,31 @@
...
 class OpenTelemetryPluginBuilderImpl;
-class OpenTelemetryPlugin;
 }  // namespace internal
 
 class OpenTelemetryPluginOption {
  public:
   virtual ~OpenTelemetryPluginOption() = default;
 };
 
+class OpenTelemetryPlugin {
```
Comment (yashykt) [inline include/grpcpp/ext/otel_plugin.h:48]: I'm thinking - mark this as an experimental API first. After we confirm that our users are happy with this, we can stabilize it.

File: include/grpcpp/ext/otel_plugin.h [hunk 5]
```diff
@@ -142,6 +149,10 @@ class OpenTelemetryPluginBuilder {
   /// Registers a global plugin that acts on all channels and servers running on
   /// the process.
   absl::Status BuildAndRegisterGlobal();
+  /// Builds an open telemetry plugin, returns the plugin object when succeeded
+  /// or an error status when failed.
+  GRPC_MUST_USE_RESULT absl::StatusOr<std::shared_ptr<OpenTelemetryPlugin>>
+  Build();
```
Comment (yashykt) [inline include/grpcpp/ext/otel_plugin.h:155]: I'm thinking - mark this as an experimental API first. After we confirm that our users are happy with this, we can stabilize it.

Comment (yashykt) [reply include/grpcpp/ext/otel_plugin.h:155]: We should also add a comment saying that you can only Build once (irrespective  of whether `Build()` is used or `BuildAndRegisterGlobal()` is used

File: include/grpcpp/ext/otel_plugin.h (hunk 2 above)
Comment (yashykt) [reply include/grpcpp/ext/otel_plugin.h:51]: I was thinking - 
`grpc::OpenTelemetryPlugin` plugin be a non-abstract class with a private member `std::shared_ptr<grpc::internal::OpenTelemetry> plugin_`.

File: include/grpcpp/ext/otel_plugin.h [hunk 6]
```diff
@@ -26,24 +26,31 @@
...
-class OpenTelemetryPlugin;
 }  // namespace internal
 
 class OpenTelemetryPluginOption {
  public:
   virtual ~OpenTelemetryPluginOption() = default;
 };
 
+class OpenTelemetryPlugin {
+ public:
```
Comment (yashykt) [inline include/grpcpp/ext/otel_plugin.h:49]: disable copy constructor

File: include/grpcpp/ext/otel_plugin.h (hunk 3 above)
Comment (yijiem) [reply include/grpcpp/ext/otel_plugin.h]: done.

File: src/core/lib/surface/legacy_channel.cc [hunk 7]
```diff
@@ -97,9 +97,23 @@ absl::StatusOr<OrphanablePtr<Channel>> LegacyChannel::Create(
...
+        GlobalStatsPluginRegistry::GetStatsPluginsForChannel(scope);
+    // Add per-channel stats plugins.
+    auto* stats_plugin_list = args.GetPointer<
+        std::shared_ptr<std::vector<std::shared_ptr<StatsPlugin>>>>(
+        GRPC_ARG_STATS_PLUGINS);
+    if (stats_plugin_list != nullptr) {
+      for (const auto& plugin : **stats_plugin_list) {
+        bool is_enabled = false;
+        std::shared_ptr<StatsPlugin::ScopeConfig> config;
+        std::tie(is_enabled, config) = plugin->IsEnabledForChannel(scope);
```
Comment (markdroth) [inline src/core/lib/surface/legacy_channel.cc:111]: This doesn't seem like quite the right interface here.  In this case, we are being explicitly told to use this stats plugin for this channel, so we shouldn't ask the plugin if it's okay and give it the opportunity to say no.  Instead, I think we should have a method on the plugin to get the config for the channel unconditionally.

File: src/cpp/ext/otel/otel_plugin.cc [hunk 8]
```diff
@@ -1007,4 +1037,9 @@ absl::Status OpenTelemetryPluginBuilder::BuildAndRegisterGlobal() {
   return impl_->BuildAndRegisterGlobal();
 }
 
+absl::StatusOr<std::shared_ptr<grpc::OpenTelemetryPlugin>>
+OpenTelemetryPluginBuilder::Build() {
+  return impl_->Build();
```
Comment (markdroth) [inline src/cpp/ext/otel/otel_plugin.cc:1042]: Not directly related to this PR, but is there anything left stopping us from moving the functionality of `OpenTelemetryPluginBuilderImpl` directly into `OpenTelemetryPluginBuilder`, so that we can stop having this level of indirection?

File: test/cpp/ext/otel/otel_test_library.h [hunk 9]
```diff
@@ -59,7 +59,7 @@ class MockMetricReader : public opentelemetry::sdk::metrics::MetricReader {
 };
 
 class OpenTelemetryPluginEnd2EndTest : public ::testing::Test {
- protected:
+ public:
```
Comment (markdroth) [inline test/cpp/ext/otel/otel_test_library.h:62]: Why does this need to be public instead of protected?  I don't think this should be used by anything but subclasses.

Is this because of the standalone `ConfigureOTBuilder()` function you have in the .cc file?  If so, I suggest making that a static method on the class instead of a standalone function.

File: test/cpp/ext/otel/otel_plugin_test.cc [hunk 10]
```diff
@@ -1969,6 +1970,92 @@ TEST(OpenTelemetryPluginMetricsEnablingDisablingTest, TestEnableDisableAPIs) {
               ::testing::UnorderedElementsAre("grpc.test.metric_3"));
 }
 
+TEST_F(OpenTelemetryPluginEnd2EndTest, RegisterMultiplePerChannelStatsPlugins) {
```
Comment (markdroth) [inline test/cpp/ext/otel/otel_plugin_test.cc:1973]: Please add tests for the following cases:
- The same stats plugin being added directly to two different channels.
- Stats plugins being added directly to servers.

File: src/core/lib/surface/legacy_channel.cc [hunk 11]
```diff
@@ -97,9 +97,23 @@ absl::StatusOr<OrphanablePtr<Channel>> LegacyChannel::Create(
...
+        if (is_enabled) {
+          (*r)->stats_plugin_group->AddStatsPlugin(plugin, std::move(config));
```
Comment (markdroth) [inline src/core/lib/surface/legacy_channel.cc:113]: @yashykt, didn't we say that we would need to store a bit in the `StatsPluginGroup` indicating whether each plugin has come from the global registry or was directly added to the channel, since we'll need that information to properly propagate stats plugins to child channels?

**Review by markdroth — COMMENTED** [review]
This looks like a great start!

Please let me know if you have any questions.  Thanks!

File: src/core/lib/surface/legacy_channel.cc (hunk 11 above)
Comment (yashykt) [reply src/core/lib/surface/legacy_channel.cc:113]: We shouldn't need to, since we already have the channel arg that only has the per channel stats plugins.

File: include/grpcpp/ext/otel_plugin.h (hunk 1 above)
Comment (yashykt) [reply include/grpcpp/ext/otel_plugin.h:154]: In the approach I was imagining, the returned object would have the `shared_ptr` member.

File: include/grpcpp/ext/otel_plugin.h (hunk 2 above)
Comment (yashykt) [reply include/grpcpp/ext/otel_plugin.h:51]: > Yash, why is that better than the virtual method?

Yeah, we talked about this offline later, and I'm fine with the current approach as well. 
+1 to renaming `internal::OpenTelemetryPlugin` if sticking with the current approach.

File: test/cpp/ext/otel/otel_plugin_test.cc (hunk 10 above)
Comment (yijiem) [reply test/cpp/ext/otel/otel_plugin_test.cc]: great ideas! done, ptal

File: test/cpp/ext/otel/otel_test_library.h (hunk 9 above)
Comment (yijiem) [reply test/cpp/ext/otel/otel_test_library.h]: exactly, done

File: include/grpcpp/ext/otel_plugin.h (hunk 4 above)
Comment (yijiem) [reply include/grpcpp/ext/otel_plugin.h:48]: done

[the remark this answers is not in the document]
Comment (yijiem) [reply include/grpcpp/ext/otel_plugin.h:48]: done

**Review by yijiem — COMMENTED** [review]
Thanks for the review!

File: BUILD [hunk 12]
```diff
@@ -2988,6 +2988,7 @@ grpc_cc_library(
     ],
     language = "c++",
     deps = [
+        ":grpc++",
```
[the remark this answers is not in the document]
Comment (markdroth) [reply BUILD]: Since we do need the impl to build, I think it's better to have the dep on `grpc++` than to use forward declarations.

**Review by markdroth — APPROVED** [review]
Looks great!  Only a few minor comments remaining.  Feel free to merge after addressing.

File: include/grpcpp/ext/otel_plugin.h [hunk 13]
```diff
@@ -146,11 +152,13 @@ class OpenTelemetryPluginBuilder {
   OpenTelemetryPluginBuilder& SetChannelScopeFilter(
       absl::AnyInvocable<bool(const ChannelScope& /*scope*/) const>
           channel_scope_filter);
-  /// Registers a global plugin that acts on all channels and servers running on
-  /// the process.
+  /// Builds and registers a global plugin that acts on all channels and servers
+  /// running on the process. Note that calling BuildAndRegisterGlobal more than
+  /// once when the first call succeeded is undefined behavior.
```
[the remark this answers is not in the document]
Comment (yijiem) [reply include/grpcpp/ext/otel_plugin.h]: done

File: include/grpcpp/ext/otel_plugin.h [hunk 14]
```diff
@@ -146,11 +152,13 @@ class OpenTelemetryPluginBuilder {
...
   absl::Status BuildAndRegisterGlobal();
   /// Builds an open telemetry plugin, returns the plugin object when succeeded
-  /// or an error status when failed.
+  /// or an error status when failed. Note that calling Build more than once
+  /// when the first call succeeded is undefined behavior.
```
[the remark this answers is not in the document]
Comment (yijiem) [reply include/grpcpp/ext/otel_plugin.h]: done

File: BUILD (hunk 12 above)
[the remark this answers is not in the document]
Comment (yijiem) [reply BUILD:2992]: done

**Review by yijiem — COMMENTED** [review]
Thanks Mark!

**Review by markdroth — APPROVED** [review]

File: include/grpcpp/ext/otel_plugin.h (hunk 5 above)
Comment (yashykt) [reply include/grpcpp/ext/otel_plugin.h:163]: Can we mark this as experimental as well please?

File: src/core/lib/channel/context.h [hunk 15]
```diff
@@ -28,9 +28,7 @@
 /// Call context is represented as an array of \a grpc_call_context_elements.
 /// This enum represents the indexes into the array, where each index
 /// contains a different type of value.
-typedef enum {
-  GRPC_CONTEXT_COUNT
-} grpc_context_index;
+typedef enum { GRPC_CONTEXT_COUNT } grpc_context_index;
```
Comment (yashykt) [inline src/core/lib/channel/context.h:31]: revert unnecessary changes

**Review by yashykt — COMMENTED** [review]
minor comments

File: include/grpcpp/ext/otel_plugin.h (hunk 5 above)
Comment (yijiem) [reply include/grpcpp/ext/otel_plugin.h:163]: done

File: include/grpcpp/ext/otel_plugin.h (hunk 4 above)
Comment (yijiem) [reply include/grpcpp/ext/otel_plugin.h:48]: done

File: src/core/lib/channel/context.h (hunk 15 above)
Comment (yijiem) [reply src/core/lib/channel/context.h:31]: done

File: include/grpcpp/ext/otel_plugin.h (hunk 4 above)
Comment (markdroth) [reply include/grpcpp/ext/otel_plugin.h:49]: If this is experimental, we should probably put it in an `experimental` namespace.

Comment (yijiem) [reply include/grpcpp/ext/otel_plugin.h:50]: done

**Review by yashykt — APPROVED** [review]
